### PR TITLE
Prioritise block hash verification error

### PR DESCRIPTION
**Pull request by jelilat**
Fixes https://github.com/NethermindEth/juno/issues/711

### CODE REVIEW

File: core/block.go [hunk 1]
```diff
@@ -101,10 +100,20 @@ func VerifyBlockHash(b *Block, network utils.Network) error {
...
 
+	if err := verifyBlockHash(b, network); err != nil {
+		return err
+	}
+
 	if err := verifyTransactions(b.Transactions, network); err != nil {
 		return err
 	}
 
+	return nil
```
Comment (omerfirmak) [inline core/block.go:111]:
```suggestion
	return verifyTransactions(b.Transactions, network)
```

**Review by omerfirmak — APPROVED** [review]
Nice catch!

File: core/block.go (hunk 1 above)
Comment (jelilat) [reply core/block.go]: Thanks for this. Fixed it

**Comment by IronGauntlets** [comment]
Checking transaction hashes before block hash was intentional because if there is an error in transaction hash calculation then the block's hash is invalid. Block hash calculation requires creating transaction commitment which requires transaction hashes.

By checking the transaction hash first and if there is an error we don't need to create a transaction commitment since we know the block hash must be incorrect.

I think what we had before was correct assuming we don't ignore transaction verification errors. 

Now if the caller of VerifyBlockHash chooses to ignore transaction verification errors there is not much we can do. We should not change this piece of code because of what the caller has decided to do.

**Comment by jelilat** [comment]
> Checking transaction hashes before block hash was intentional because if there is an error in transaction hash calculation then the block's hash is invalid. Block hash calculation requires creating transaction commitment which requires transaction hashes.
> 
> By checking the transaction hash first and if there is an error we don't need to create a transaction commitment since we know the block hash must be incorrect.
> 
> I think what we had before was correct assuming we don't ignore transaction verification errors.
> 
> Now if the caller of VerifyBlockHash chooses to ignore transaction verification errors there is not much we can do. We should not change this piece of code because of what the caller has decided to do.

Got you. But in this case, it wasn't the caller who decided that they didn't want to verify blocks. Rather, we are the ones who isn't verifying the block.

Is it safe to say that we aren't actually verifying blocks then?

**Comment by jelilat** [comment]
> Checking transaction hashes before block hash was intentional because if there is an error in transaction hash calculation then the block's hash is invalid. Block hash calculation requires creating transaction commitment which requires transaction hashes.
> 
> By checking the transaction hash first and if there is an error we don't need to create a transaction commitment since we know the block hash must be incorrect.
> 
> I think what we had before was correct assuming we don't ignore transaction verification errors.
> 
> Now if the caller of VerifyBlockHash chooses to ignore transaction verification errors there is not much we can do. We should not change this piece of code because of what the caller has decided to do.

Got you

**Comment by IronGauntlets** [comment]
> Rather, we are the ones who isn't verifying the block.

Yes, but we need to keep in mind the separation of concerns. We need to write a function in a manner which makes sense as to what it is trying to achieve.

> Is it safe to say that we aren't actually verifying blocks then?

This is not true, we are verifying the blocks but our sync package has chosen to make progress despite some errors. Not verifying them would be to not even acknowledge that there is a failure.

Also, after regenesis when we don't ignore errors we will have to revert this change.

**Pull request closed without merging**